### PR TITLE
Add KeyVault.Secrets configuration extensions

### PR DESCRIPTION
**Pull request by pakrym**
Add an initial version of KeyVault integration package for Microsoft.Extensions.Configuration.

It's planned to re-review the API but I want to get it building.

### CODE REVIEW

File: sdk/keyvault/Azure.Security.KeyVault.Secrets.Extensions.Configuration/api/Azure.Security.KeyVault.Secrets.Extensions.Configuration.netstandard2.0.cs [hunk 1]
```diff
@@ -0,0 +1,34 @@
+namespace Azure.Security.KeyVault.Secrets.Extensions.Configuration
+{
+    public partial class AzureKeyVaultConfigurationOptions
+    {
+        public AzureKeyVaultConfigurationOptions() { }
+        public AzureKeyVaultConfigurationOptions(System.Uri vault, Azure.Core.TokenCredential credential) { }
```
Comment (heaths) [inline sdk/keyvault/Azure.Security.KeyVault.Secrets.Extensions.Configuration/api/Azure.Security.KeyVault.Secrets.Extensions.Configuration.netstandard2.0.cs]: We use `vaultUri` in Key Vault, but if this is convention in similar use patterns, so be it.

File: sdk/keyvault/Azure.Security.KeyVault.Secrets.Extensions.Configuration/api/Azure.Security.KeyVault.Secrets.Extensions.Configuration.netstandard2.0.cs [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+        bool Load(Azure.Security.KeyVault.Secrets.SecretProperties secret);
+    }
+}
+namespace Microsoft.Extensions.Configuration
+{
+    public static partial class AzureKeyVaultConfigurationExtensions
+    {
+        public static Microsoft.Extensions.Configuration.IConfigurationBuilder AddAzureKeyVault(this Microsoft.Extensions.Configuration.IConfigurationBuilder configurationBuilder, Azure.Security.KeyVault.Secrets.Extensions.Configuration.AzureKeyVaultConfigurationOptions options) { throw null; }
+        public static Microsoft.Extensions.Configuration.IConfigurationBuilder AddAzureKeyVault(this Microsoft.Extensions.Configuration.IConfigurationBuilder configurationBuilder, Azure.Security.KeyVault.Secrets.SecretClient client, Azure.Security.KeyVault.Secrets.Extensions.Configuration.IKeyVaultSecretManager manager) { throw null; }
+        public static Microsoft.Extensions.Configuration.IConfigurationBuilder AddAzureKeyVault(this Microsoft.Extensions.Configuration.IConfigurationBuilder configurationBuilder, System.Uri vault) { throw null; }
```
Comment (heaths) [inline sdk/keyvault/Azure.Security.KeyVault.Secrets.Extensions.Configuration/api/Azure.Security.KeyVault.Secrets.Extensions.Configuration.netstandard2.0.cs]: Same comment here and below for `vaultUri`.

File: sdk/keyvault/Azure.Security.KeyVault.Secrets.Extensions.Configuration/src/Azure.Security.KeyVault.Secrets.Extensions.Configuration.csproj [hunk 3]
```diff
@@ -0,0 +1,17 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <Description>Azure KeyVault configuration provider implementation for Microsoft.Extensions.Configuration.</Description>
```
Comment (heaths) [inline sdk/keyvault/Azure.Security.KeyVault.Secrets.Extensions.Configuration/src/Azure.Security.KeyVault.Secrets.Extensions.Configuration.csproj]: "Key Vault" when used in writing. Check the other PR you did at the same time (I might've missed it).

File: sdk/keyvault/Azure.Security.KeyVault.Secrets.Extensions.Configuration/src/AzureKeyVaultConfigurationExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+namespace Microsoft.Extensions.Configuration
+#pragma warning restore
+{
+    /// <summary>
+    /// Extension methods for registering <see cref="AzureKeyVaultConfigurationProvider"/> with <see cref="IConfigurationBuilder"/>.
+    /// </summary>
+    public static class AzureKeyVaultConfigurationExtensions
+    {
+        /// <summary>
+        /// Adds an <see cref="IConfigurationProvider"/> that reads configuration values from the Azure KeyVault.
```
Comment (heaths) [inline sdk/keyvault/Azure.Security.KeyVault.Secrets.Extensions.Configuration/src/AzureKeyVaultConfigurationExtensions.cs]: "Key Vault" in writing.

Comment (heaths) [reply sdk/keyvault/Azure.Security.KeyVault.Secrets.Extensions.Configuration/src/AzureKeyVaultConfigurationExtensions.cs]: (I'll not repeat, so please check for other instances.)

File: sdk/keyvault/Azure.Security.KeyVault.Secrets.Extensions.Configuration/src/AzureKeyVaultConfigurationExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+
+        /// <summary>
+        /// Adds an <see cref="IConfigurationProvider"/> that reads configuration values from the Azure KeyVault.
+        /// </summary>
+        /// <param name="configurationBuilder">The <see cref="IConfigurationBuilder"/> to add to.</param>
+        /// <param name="options">The <see cref="AzureKeyVaultConfigurationOptions"/> to use.</param>
+        /// <returns>The <see cref="IConfigurationBuilder"/>.</returns>
+        public static IConfigurationBuilder AddAzureKeyVault(this IConfigurationBuilder configurationBuilder, AzureKeyVaultConfigurationOptions options)
+        {
+            if (configurationBuilder == null)
```
Comment (heaths) [inline sdk/keyvault/Azure.Security.KeyVault.Secrets.Extensions.Configuration/src/AzureKeyVaultConfigurationExtensions.cs]: Use the `Argument` shared class for consistency. Add a ProjectReference to the shared project.

File: sdk/keyvault/Azure.Security.KeyVault.Secrets.Extensions.Configuration/src/AzureKeyVaultConfigurationProvider.cs [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+
+        /// <summary>
+        /// Creates a new instance of <see cref="AzureKeyVaultConfigurationProvider"/>.
+        /// </summary>
+        /// <param name="client">The <see cref="SecretClient"/> to use for retrieving values.</param>
+        /// <param name="manager">The <see cref="IKeyVaultSecretManager"/> to use in managing values.</param>
+        /// <param name="reloadInterval">The timespan to wait in between each attempt at polling the Azure KeyVault for changes. Default is null which indicates no reloading.</param>
+        public AzureKeyVaultConfigurationProvider(SecretClient client, IKeyVaultSecretManager manager, TimeSpan? reloadInterval = null)
+        {
+            _client = client ?? throw new ArgumentNullException(nameof(client));
```
Comment (heaths) [inline sdk/keyvault/Azure.Security.KeyVault.Secrets.Extensions.Configuration/src/AzureKeyVaultConfigurationProvider.cs]: Use the `Argument` class for consistency.

File: sdk/keyvault/Azure.Security.KeyVault.Secrets.Extensions.Configuration/src/AzureKeyVaultConfigurationProvider.cs [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+            while (!_cancellationToken.IsCancellationRequested)
+            {
+                await WaitForReload().ConfigureAwait(false);
+                try
+                {
+                    await LoadAsync().ConfigureAwait(false);
+                }
+                catch (Exception)
+                {
+                    // Ignore
```
Comment (heaths) [inline sdk/keyvault/Azure.Security.KeyVault.Secrets.Extensions.Configuration/src/AzureKeyVaultConfigurationProvider.cs]: No logging or tracing at least?

File: sdk/keyvault/Azure.Security.KeyVault.Secrets.Extensions.Configuration/src/AzureKeyVaultConfigurationProvider.cs [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+            {
+                var secretBundle = task.Result;
+                newLoadedSecrets.Add(secretBundle.Value.Name, new LoadedSecret(_manager.GetKey(secretBundle), secretBundle.Value.Value, secretBundle.Value.Properties.UpdatedOn));
+            }
+
+            _loadedSecrets = newLoadedSecrets;
+
+            // Reload is needed if we are loading secrets that were not loaded before or
+            // secret that was loaded previously is not available anymore
+            if (tasks.Any() || oldLoadedSecrets?.Any() == true)
```
Comment (heaths) [inline sdk/keyvault/Azure.Security.KeyVault.Secrets.Extensions.Configuration/src/AzureKeyVaultConfigurationProvider.cs]: This boxes `true`. Check instead if `oldLoadedScrets` is null then check `Any()`. I also wrote an `IsNullOrEmpty()` extension method that may be in scope if you reference our shared project.

File: sdk/keyvault/Azure.Security.KeyVault.Secrets.Extensions.Configuration/src/AzureKeyVaultConfigurationProvider.cs [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+            public DateTimeOffset? Updated { get; }
+
+            public bool IsUpToDate(DateTimeOffset? updated)
+            {
+                if (updated.HasValue != Updated.HasValue)
+                {
+                    return false;
+                }
+
+                return updated.GetValueOrDefault() == Updated.GetValueOrDefault();
```
Comment (heaths) [inline sdk/keyvault/Azure.Security.KeyVault.Secrets.Extensions.Configuration/src/AzureKeyVaultConfigurationProvider.cs]: I'm not understanding why this is simple equality. Care to clarify?

**Review by heaths — APPROVED** [review]

File: sdk/keyvault/Azure.Security.KeyVault.Secrets.Extensions.Configuration/src/AzureKeyVaultConfigurationProvider.cs (hunk 8 above)
Comment (pakrym) [reply sdk/keyvault/Azure.Security.KeyVault.Secrets.Extensions.Configuration/src/AzureKeyVaultConfigurationProvider.cs]: true is not being boxed here, `oldLoadedSecrets?.Any() == true` is compiled to `list == null || !Enumerable.Any(list)`

File: sdk/keyvault/Azure.Security.KeyVault.Secrets.Extensions.Configuration/src/AzureKeyVaultConfigurationProvider.cs (hunk 9 above)
Comment (pakrym) [reply sdk/keyvault/Azure.Security.KeyVault.Secrets.Extensions.Configuration/src/AzureKeyVaultConfigurationProvider.cs]: Updated value can only increase so equality is enough here, if we don't have the exact time it's not up-to-date.

File: sdk/keyvault/Azure.Security.KeyVault.Secrets.Extensions.Configuration/src/AzureKeyVaultConfigurationProvider.cs (hunk 7 above)
Comment (pakrym) [reply sdk/keyvault/Azure.Security.KeyVault.Secrets.Extensions.Configuration/src/AzureKeyVaultConfigurationProvider.cs]: `ILogger` is not available in configuration providers and we didn't seem to have a problem with this before. Usually the `Load` method throws first and surfaces the error.

**Pull request merged**